src/libcore/alloc.rs

```diff
@@ -579,6 +579,17 @@ pub unsafe trait GlobalAlloc {
     }
 }
 
+/// A hack so the default impl can condition on the associated type. This `Err`
+/// type ought to just live in the `Alloc` trait.
+#[unstable(feature = "allocator_api", issue = "32838")]
+pub trait AllocHelper {
```

Comment:
I'm fairly certain this trait needn't exist. Since `Alloc` is still experimental, we can make breaking changes.

Reply:
See https://github.com/rust-lang/rust/issues/52396 this was the only way to make it type check without removing the default methods. Happy to get rid of it once bug is fixed.

Reply:
Doesn't matter if it's a breaking change though.

Reply:
I think we need to have some default methods; not having that constrains the design a lot. It should be a simple bug to fix.

Reply:
Maybe I'm misunderstanding the issue. There are plenty of traits with associated types and default methods in the standard library. `Iterator` is an example. Don't use specialization, and it should be fine.

Reply:
The issue is when the default method is only valid for certain associated types. You can't write that constraint today in a way that works.

Reply:
If we remove the `{alloc|dealloc}_array` and `{alloc|dealloc}_one` from the `Alloc` trait like [this issue](https://github.com/rust-lang/wg-allocators/issues/18) suggests, we don't need any fancy specialization or the helper trait.

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=749a844f3b19c589228ba9ef4286b233

Reply:
That's true. I still want that bug fixed though! :D That fact that one can write `Alloc<Err = !>` means that `AllocHelper` should't need to be referenced (almost?) anywhere. We have a little runway because we need a compiler change for unstable type parameters anyway.

Reply:
It's still inelegant to have the helper trait, so I'd like to avoid it if possible.